llarp/router/rc_lookup_handler.cpp

```diff
@@ -0,0 +1,323 @@
+    {
+      for(const auto &rc : _bootstrapRCList)
+      {
+        _dht->impl->ExploreNetworkVia(dht::Key_t{rc.pubkey});
+      }
+    }
+    else
+    {
+      LogError("we have no bootstrap nodes specified");
+    }
+
+    // TODO: only explore via random subset
+    // explore via every connected peer
+    _linkManager->ForEachPeer([&](ILinkSession *s) {
+      if(!s->IsEstablished())
+        return;
+      const RouterContact rc = s->GetRemoteRC();
+      if(rc.IsPublicRouter())
+        _dht->impl->ExploreNetworkVia(dht::Key_t{rc.pubkey});
+    });
+  }
+
+  void
+  RCLookupHandler::Init(llarp_dht_context *dht,
+                        llarp_nodedb *nodedb,
+                        ILinkManager *linkManager,
+                        service::Context *hiddenServiceContext,
+                        const std::set< RouterID > &strictConnectPubkeys,
+                        const std::set< RouterContact > &bootstrapRCList,
+                        bool useWhitelist_arg,
+                        bool isServiceNode_arg)
+  {
+    _dht = dht;
+    _nodedb = nodedb;
+    _hiddenServiceContext = hiddenServiceContext;
+    _strictConnectPubkeys = strictConnectPubkeys;
+    _bootstrapRCList = bootstrapRCList;
+    _linkManager = linkManager;
+    useWhitelist = useWhitelist_arg;
+    isServiceNode = isServiceNode_arg;
+
+    for (const auto& rc : _bootstrapRCList)
+    {
+      _bootstrapRouterIDList.insert(rc.pubkey);
+    }
+  }
+
+  void
+  RCLookupHandler::HandleDHTLookupResult(RouterID remote,
+                                         const std::vector< RouterContact > &results)
+  {
+    if(not results.size())
+    {
+      FinalizeRequest(remote, nullptr, RCRequestResult::RouterNotFound);
+      return;
+    }
+
+    if (not RemoteIsAllowed(remote))
+    {
+      FinalizeRequest(remote, &results[0], RCRequestResult::InvalidRouter);
+      return;
+    }
+
+    if (not CheckRC(results[0]))
+    {
+      FinalizeRequest(remote, &results[0], RCRequestResult::BadRC);
+      return;
+    }
+
+    FinalizeRequest(remote, &results[0], RCRequestResult::Success);
+  }
+
+  bool
+  RCLookupHandler::HavePendingLookup(RouterID remote) const
+  {
+    return pendingCallbacks.find(remote) != pendingCallbacks.end();
+  }
+
+  bool
+  RCLookupHandler::RemoteInBootstrap(const RouterID& remote) const
+  {
+    for (const auto& rc : _bootstrapRCList)
+    {
+      if (rc.pubkey == remote)
+      {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void
+  RCLookupHandler::FinalizeRequest(const RouterID &router,
+                                   const RouterContact * const rc,
+                                   RCRequestResult result)
+  {
+    util::Lock l(&_mutex);
+
+    auto itr = pendingCallbacks.find(router);
+
+    if (itr != pendingCallbacks.end())
+    {
+      
+      auto& callbackQueue = itr->second;
+
+      while (not callbackQueue.empty())
+      {
+        callbackQueue.front()(router, rc, result);
```

Comment:
what happens if the callback eventually wants to acquire the RCLookupHandler lock again?

Reply:
fair point.  I don't particularly like the idea of collecting the callbacks in a local container first (and then unlocking) but that's probably for the best.

Reply:
Addressed.